crates/llms/src/bedrock/chat/mod.rs

```diff
@@ -0,0 +1,619 @@
+            .output
+            .map(|o| {
+                let Message { role, content, .. } =
+                    o.as_message().map_err(|e| to_api_error(format!("{e:?}")))?;
+
+                let data: Vec<(
+                    (Option<String>, Option<String>),
+                    Option<ChatCompletionMessageToolCall>,
+                )> = content
+                    .iter()
+                    .map(extract_from_content_block)
+                    .collect::<Result<Vec<_>, _>>()?;
+                let (content_and_refusal, tool_calls): (Vec<_>, Vec<_>) = data.into_iter().unzip();
+                let (content, refusals): (Vec<_>, Vec<_>) = content_and_refusal.into_iter().unzip();
+
+                Ok::<_, OpenAIError>(ChatChoice {
+                    index: 0,
+                    message: ChatCompletionResponseMessage {
+                        content: Some(content.into_iter().flatten().join("\n")),
+                        refusal: Some(refusals.into_iter().flatten().join("\n")),
+                        tool_calls: Some(tool_calls.into_iter().flatten().collect()),
+                        role: try_convert_role(role)?,
+                        function_call: None,
+                        audio: None,
+                    },
+                    logprobs: None,
+                    finish_reason: Some(try_convert_finish_reason(&output.stop_reason)?),
+                })
+            })
+            .transpose()?
+        else {
+            return Err(to_api_error(
+                "No outputs received from AWS bedrock converse API",
+            ));
+        };
+
+        Ok(CreateChatCompletionResponse {
+            usage,
+            id: Span::current()
+                .id()
+                .map(|id| id.into_u64().to_string())
+                .unwrap_or_default(),
+            choices: vec![choices],
+            created: SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .map_err(|e| OpenAIError::InvalidArgument(e.to_string()))?
+                .as_secs() as u32,
+            model: self.model_id.clone(),
+            service_tier: None,
+            system_fingerprint: None,
+            object: "chat.completion".to_string(),
+        })
+    }
+
+    #[allow(clippy::too_many_lines)]
+    fn process_stream(
+        model: String,
+        mut input_stream: EventReceiver<ConverseStreamOutputPacket, ConverseStreamOutputError>,
+    ) -> ChatCompletionResponseStream {
+        // Individual packets in the stream may need information from previous packets to be converted correctly to a  [`CreateChatCompletionStreamResponse`]. This struct track all the necessary information.
+        #[derive(Clone, Default)]
+        struct StreamState {
+            id: String,
+            role: Option<Role>,
+            content_block_index_to_tool_details: HashMap<i32, ToolUseBlockStart>,
+            content_block_index_to_delta_idx: HashMap<i32, u32>,
+        }
+
+        let s = stream! {
```

Comment:
Instead of using the `stream!` macro, using `unfold` here (https://docs.rs/futures/latest/futures/stream/fn.unfold.html) and keeping the input stream as part of the `StreamState` might be a more natural fit. And has the benefit of being natural Rust instead of macro Rust.

Reply:
I'm not a huge fan of large code blocks in `stream!` macros - it breaks code completion for rust-analyzer

Reply:
Curious what those issues were? I'm not sure if a map will work here with the stream state - it isn't a pure function of the inputs

Reply:
wouldn't you just keep looping then in the unfold implementation until you can emit an item?